### ISSUE TITLE
KD stage1 train loss Nan

### ISSUE THREAD

**Issue by ztwe** [Yochengliu/MLIC-KD-WSD#9]
pretrain model: VGG_ILSVRC_16_layers.caffemodel

solver:
```
net: "kd/train_stage1.prototxt"
test_iter: 800
test_interval: 2400
base_lr: 0.00001
momentum: 0.9
type: "SGD"
weight_decay: 0.0005
lr_policy: "inv"
gamma: 0.00001
power: 0.75
display: 200
max_iter: 800000
snapshot: 60000
snapshot_prefix: "models"
solver_mode: GPU
```

![image](https://user-images.githubusercontent.com/7899459/49682905-1e7a5580-faf7-11e8-8935-05feaab9b3eb.png)

**Comment by ztwe** [comment]
If pretrain model changes to WSDDN, both train loss and test loss drop quite quickly. The test loss is about 0.07 after 10000 iter. The train loss is about 0.00x and sometimes raises up 2.x. Is that right?



if loading two pretrain models using 
'--weights=modesl_iter_100000.caffemodel,VGG_ILSVRC_16_layers.caffemodel', the loss is also Nan.

**Comment by Yochengliu** [comment]
Because the names of caffe layers between the T-WDet and S-Cls are different in stage1 prototxt, by using  '--weights=modesl_iter_100000.caffemodel,VGG_ILSVRC_16_layers.caffemodel', the weight of T-WDet is actually not initialized with modesl_iter_100000.caffemodel. You can check this by comparing  WSDDN prototxt and stage1 prototxt. 

This issue can be fixed by modifying the layer names in modesl_iter_100000.caffemodel to be consistent with the ones in stage1 prototxt.

**Comment by ztwe** [comment]
On saturday I realized the layer names may be a problem when restored the weights.
But how to modify the layer names?
I'm not familiar with caffe while I have been used to develop with tf.

Is the following code right?

```
import sys
from caffe import Net
n_orig = Net('test_o.prototxt', 'orig.caffemodel', 0)
n_new = Net('test_n.prototxt', 'orig.caffemodel', 0)
n_new.params['newlayername'] = n_orig.params['origlayername']
n_new.save('new.caffemodel')
```

**Comment by ztwe** [comment]
I try the following code:

```
import sys
from caffe import Net
n_orig = Net('wsddn_train.prototxt', '../models_iter_500000.caffemodel', 0)
n_new = Net('wsddn_train_for_stage1.prototxt', '../models_iter_500000.caffemodel', 0)
n_new.params['conv_wsd_1_1'] = n_orig.params['conv1_1']
n_new.params['conv_wsd_1_2'] = n_orig.params['conv1_2']
n_new.params['pool_wsd_1'] = n_orig.params['pool1']
n_new.params['conv_wsd_2_1'] = n_orig.params['conv2_1']
n_new.params['conv_wsd_2_2'] = n_orig.params['conv2_2']
n_new.params['pool_wsd_2'] = n_orig.params['pool2']
n_new.params['conv_wsd_3_1'] = n_orig.params['conv3_1']
n_new.params['conv_wsd_3_2'] = n_orig.params['conv3_2']
n_new.params['conv_wsd_3_3'] = n_orig.params['conv3_3']
n_new.params['pool_wsd_3'] = n_orig.params['pool3']
n_new.params['conv_wsd_4_1'] = n_orig.params['conv4_1']
n_new.params['conv_wsd_4_2'] = n_orig.params['conv4_2']
n_new.params['conv_wsd_4_3'] = n_orig.params['conv4_3']
n_new.params['pool_wsd_4'] = n_orig.params['pool4']
n_new.params['conv_wsd_5_1'] = n_orig.params['conv5_1']
n_new.params['conv_wsd_5_2'] = n_orig.params['conv5_2']
n_new.params['conv_wsd_5_3'] = n_orig.params['conv5_3']
n_new.params['fc_wsd_6'] = n_orig.params['fc6']
n_new.params['fc_wsd_7'] = n_orig.params['fc7']

n_new.save('wsddn_for_stage1.caffemodel')
```

When transforming, error occurs.

![image](https://user-images.githubusercontent.com/7899459/49721492-44d9f580-fc9d-11e8-8392-2aa7c4de23af.png)

**Comment by ztwe** [comment]
I transform the origin wsddn model into wsddn_for_stage1.caffemodel using the above code.

when starting training using the following command with both wsddn_for_stage1.caffemodel and VGG-16.caffemodel , the loss is Nan.
![image](https://user-images.githubusercontent.com/7899459/49723012-e9116b80-fca0-11e8-8db1-41102f99deb7.png)


when starting training using the following command with only wsddn_for_stage1.caffemodel,  the loss is 0.
![image](https://user-images.githubusercontent.com/7899459/49722953-ca12d980-fca0-11e8-8be4-98b2f2f2ea30.png)


![image](https://user-images.githubusercontent.com/7899459/49722692-4527c000-fca0-11e8-98d4-489c87daf9d2.png)


It seems the conversion between the origin wsddn model and wsddn_for_stage1.caffemodel is fails?

**Comment by ztwe** [comment]
import sys
from caffe import Net
n_orig = Net('wsddn_train_origin.prototxt', '../models_iter_50000.caffemodel', 0)
n_new = Net('wsddn_train_for_stage1.prototxt', '../models_iter_50000.caffemodel', 0)
n_new.params['conv_wsd_1_1'] = n_orig.params['conv1_1']
n_new.params['conv_wsd_1_2'] = n_orig.params['conv1_2']
#n_new.params['pool_wsd_1'] = n_orig.params['pool1']
n_new.params['conv_wsd_2_1'] = n_orig.params['**_conv2_3_**']
n_new.params['**_conv_wsd_2_3_**'] = n_orig.params['conv2_2']
#n_new.params['pool_wsd_2'] = n_orig.params['pool2']
n_new.params['conv_wsd_3_1'] = n_orig.params['conv3_1']
n_new.params['conv_wsd_3_2'] = n_orig.params['conv3_2']
n_new.params['conv_wsd_3_3'] = n_orig.params['conv3_3']
#n_new.params['pool_wsd_3'] = n_orig.params['pool3']
n_new.params['conv_wsd_4_1'] = n_orig.params['conv4_1']
n_new.params['conv_wsd_4_2'] = n_orig.params['conv4_2']
n_new.params['conv_wsd_4_3'] = n_orig.params['conv4_3']
#n_new.params['pool_wsd_4'] = n_orig.params['pool4']
n_new.params['conv_wsd_5_1'] = n_orig.params['conv5_1']
n_new.params['conv_wsd_5_2'] = n_orig.params['conv5_2']
n_new.params['conv_wsd_5_3'] = n_orig.params['conv5_3']
n_new.params['fc_wsd_6'] = n_orig.params['fc6']
n_new.params['fc_wsd_7'] = n_orig.params['fc7']

n_new.save('wsddn_for_stage1.caffemodel')

**Comment by ztwe** [comment]
During transformation, all source layers have been ignored. Why?

![image](https://user-images.githubusercontent.com/7899459/49725450-a9e61900-fca6-11e8-823a-9d704dae8eab.png)

**Comment by ztwe** [comment]
I look into the net.cpp.

```
    const LayerParameter& source_layer = param.layer(i);
    const string& source_layer_name = source_layer.name();
    int target_layer_id = 0;
    while (target_layer_id != layer_names_.size() &&
        layer_names_[target_layer_id] != source_layer_name) {
      ++target_layer_id;
    }
    if (target_layer_id == layer_names_.size()) {
      LOG(INFO) << "Ignoring source layer " << source_layer_name;
      continue;
    }

```

**Comment by ztwe** [comment]
I replace 
`n_new.params['xxx'] = n_orig.params['yyy']`
with
n_new.params['xxx'][0].data[...] = n_orig.params['yyy'][0].data[...]
n_new.params['xxx'][1].data[...] = n_orig.params['yyy'][1].data[...]

It seems to be right.
Now the loss of stage1 is shown as following:

![image](https://user-images.githubusercontent.com/7899459/49733843-8f6b6a00-fcbd-11e8-8682-94384a6532a6.png)

The initial test loss is 5000. It decreases to 2200 when iter=2400.

Is that right?

**Comment by Yochengliu** [comment]
It seems that there is no "conv2_3".

In terms of early training shown in the screenshot, it looks all right.

**Comment by ztwe** [comment]
Yes, "conv2_3" is just used to check whether the  transformation is really implemented.

what are the final stage1 loss and stage2 loss in your training?

**Comment by Yochengliu** [comment]
The training loss of s1 is about 3800 (no test loss, we didn't test the model in s1). The loss_sig and loss_soft of s2 are both about 4.

**Comment by Yochengliu** [comment]
To be specific, the training loss in s2 is about 6. The test loss_sig and loss_soft of s2 are both about 4.

**Comment by ztwe** [comment]
train_stage1.prototxt contains the content shown as following.

```
layer{
  name: "data"
  type: "HumanAttData"
  top: "data"
  top: "label_roi"
  include {
    phase: TEST
  }
  transform_param {
    mirror: false
    mean_value: 103.52
    mean_value: 116.28
    mean_value: 123.675
  }
  human_att_data_param {
    source: "test_672_mscoco_edge_box_score.txt"
    root_folder: ""
    batch_size: 16
    shuffle: false
    new_width: 672
    new_height: 672
    att_num: 80
    bbox_num: 500
  }
}
```
I have to focus on the test loss because train loss varies dramatically while batch_size=3.

![image](https://user-images.githubusercontent.com/7899459/49871959-95468400-fe52-11e8-8c52-12dd7cca5f40.png)
